crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,248 @@
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_UINT64 as u64;
+pub const NVTX_PAYLOAD_ENTRY_TYPE_COLOR_ARGB: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_COLOR_ARGB as u64;
+pub const NVTX_PAYLOAD_ENTRY_TYPE_CSTRING: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_ENTRY_TYPE_CSTRING as u64;
+pub const NVTX_PAYLOAD_SCHEMA_TYPE_STATIC: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_TYPE_STATIC as u64;
+pub const NVTX_PAYLOAD_SCHEMA_TYPE_DYNAMIC: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_TYPE_DYNAMIC as u64;
+pub const NVTX_PAYLOAD_SCHEMA_FLAG_NONE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_FLAG_NONE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_FLAG_REFERENCED: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_FLAG_REFERENCED as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NAME: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NAME as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_TYPE: u64 = nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_TYPE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_FLAGS as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_NUM_ENTRIES as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_STATIC_SIZE as u64;
+pub const NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID: u64 =
+    nvtxw_bindings::NVTX_PAYLOAD_SCHEMA_ATTR_SCHEMA_ID as u64;
+
+pub const NVTXW3_STREAM_ORDER_INTERLEAVING_NONE: i16 = 0;
+pub const NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN: i16 = 0;
+
+fn check(result: ResultCode) -> Result<(), ResultCode> {
+    if result != NVTXW3_RESULT_SUCCESS {
+        return Err(result);
+    }
+    Ok(())
+}
+
+pub fn initialize_simple() -> Result<InterfaceHandle, ResultCode> {
+    let so_name = "libNvtxwBackend.so";
+    let c_mode_string = CString::new(so_name).expect("modeString CString::new failed");
+    let cptr_mode_string = c_mode_string.as_ptr();
+
+    let mut get_interface_func: nvtxw_bindings::nvtxwGetInterface_t = Default::default();
+    check(unsafe {
+        nvtxw_bindings::nvtxwInitialize(
+            NVTXW3_INIT_MODE_LIBRARY_FILENAME,
+            cptr_mode_string,
+            &mut get_interface_func,
+            null_mut(),
+        )
+    })?;
+
+    let get_interface = get_interface_func.ok_or(NVTXW3_RESULT_FAILED)?;
+
+    let mut interface_void: *const c_void = null();
+    check(unsafe { get_interface(NVTXW3_INTERFACE_ID_CORE_V1, &mut interface_void) })?;
+
+    if interface_void.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    let ptr_interface = interface_void as *const nvtxw_bindings::nvtxwInterfaceCore_t;
+    let interface;
+    unsafe {
+        interface = *ptr_interface;
+    }
+
+    Ok(interface)
+}
+
+pub fn session_begin_simple(
+    interface: InterfaceHandle,
+    o_output: Option<PathBuf>,
+    o_merge: Option<PathBuf>,
```

Comment:
Generally speaking it's not typical to see Rust code encoding the types into the names of variables. Usually if you are going through some sort of a multi-step type conversion process, you chain the conversion operators like I suggest below.

Reply:
Fixed